uniffi_bindgen/src/bindings/kotlin/gen_kotlin/wrapped.rs

```diff
@@ -41,20 +63,52 @@ impl CodeType for WrappedCodeType {
         nm: &dyn fmt::Display,
         target: &dyn fmt::Display,
     ) -> String {
-        self.prim.write(oracle, nm, target)
+        format!(
+            "{}.write({}, {})",
+            self.ffi_converter_name(oracle),
+            oracle.var_name(nm),
+            target
+        )
     }
 
     fn lift(&self, oracle: &dyn CodeOracle, nm: &dyn fmt::Display) -> String {
-        self.prim.lift(oracle, nm)
+        format!("{}.lift({})", self.ffi_converter_name(oracle), nm)
     }
 
     fn read(&self, oracle: &dyn CodeOracle, nm: &dyn fmt::Display) -> String {
-        self.prim.read(oracle, nm)
+        format!("{}.read({})", self.ffi_converter_name(oracle), nm)
+    }
+
+    fn imports(&self, _oracle: &dyn CodeOracle) -> Option<Vec<String>> {
+        match &self.config {
+            None => None,
+            Some(wrapper_config) => wrapper_config.imports.clone(),
+        }
     }
 
     fn helper_code(&self, _oracle: &dyn CodeOracle) -> Option<String> {
-        // We don't want the helper code for the wrapped type, otherwise we
-        // will end up with 2 copies of it, which breaks things.
-        None
+        Some(KotlinWrappedType::new(self).render().unwrap())
+    }
+}
+
+#[derive(Template)]
+#[template(syntax = "kt", escape = "none", path = "WrappedTemplate.kt")]
+pub struct KotlinWrappedType {
+    name: String,
+    inner: Type,
+    config: Option<WrapperConfig>,
+}
+
+impl KotlinWrappedType {
+    pub fn new(codetype: &WrappedCodeType) -> Self {
+        Self {
+            name: codetype.name.clone(),
+            inner: codetype.inner.clone(),
+            config: codetype.config.clone(),
+        }
+    }
+
+    fn inner_ffi_type(&self) -> FFIType {
+        (&self.inner).into()
```

Comment:
> The KotlinWrappedType should be impl CodeDeclaration, and then that can render the declaration_code and imports instead of going through the WrappedCodeType.

Can you explain this one a bit more?

One thing I notice is that in the other modules, we split up the code to handle the `Type` vs the code to handle the definition struct.  `ObjectCodeType` handles `Type::Object` and `KotlinObject` handles `Object`.  This got tricky for wrapper types, because there's no separate definition struct -- `Type::Wrapped` contains all the data.  I'm also just not sure of why we need that split, IMO it would be nicer if `ObjectCodeType` and `KotlinObject` were merged into 1 struct.

> All the languages we support support type aliasing. If the config doesn't specify wrapping expressions, the new name should still be used, as it conveys the intent of the API design, and it will simplify/clarify this CodeType/WrappedType.
> I suspect if you do these things, then the need for a stateful oracle will go away.

I think you're saying that we could have `WrappedType` not store the config, and then make `imports()` and `helper_code()` input it as an argument.  That was my initial plan, but there's one very annoying issue there: Each language has a different `Config`, which makes adding it as an argument to the `CodeDeclaration` trait tricky.  I think there's a way to parameterize the trait, but it's not so simple.  I also thought about just using `Any` as the parameter type, but I didn't like that so much.

Reply:
> because there's no separate definition struct

Oh. Perhaps it should, it declares another type.

declare your KotlinWrappedType (arguably a `TypeDefCodeDeclaration`) as a `CodeDeclaration`. 

```rust
#[derive(Template)]
#[template(syntax = "kt", escape = "none", path = "WrappedTemplate.kt")]
pub struct KotlinWrappedType {
    name: String,
    inner: Type,
    config: Option<WrapperConfig>,
}
/// snip
impl CodeDeclaration for KotlinWrappedType {
    fn definition_code(&self, _oracle: &dyn CodeOracle) -> Option<String> {
        Some(self.render().unwrap())
    }

    fn imports(&self, _oracle: &dyn CodeOracle) -> Option<Vec<String>> {
        Some(self.config.as_ref()?.imports.as_ref()?.clone())
    }
}
```

add it to the `members()` method: 

```rust
        .chain(
            ci.iter_wrapped_types() // handwave
                .into_iter()
                .map(|inner| {
                      // handwave, get the wrapper config etc, then make
                    Box::new(wrapped::KotlinWrappedType::new(…)))
                        as Box<dyn CodeDeclaration>
                }),
        )
```

and the imports and helper code do their thing without the `CodeType` needing to have the config file.

Reply:
The general rule is the `CodeType` is used when other types need to refer to a type.

The `CodeDeclaration` is used for every type that is declared in the UDL.

So, Objects has both a `CodeType` because other types refer to it, and at least one declaration.

`String` isn't declared in the UDL, so doesn't have a `CodeDecalation`, but does have a `CodeType`.

Reply:
That makes sense and I think that would work for this change, but I like having `CodeOracle` store the `Config`.  It seems like it enables a lot of things and it's something we want going forward.  For example I want to allow the config to wrap builtin types like `timestamp` or user-defined structs without needing to define the wrapper in the UDL.  That brings up the same issue of the type name depending on the config.  We could make it so every type always uses a type alias, but seems like it makes the generate code needlessly complicated.

Reply:
Isn't this the point of so called External types? Isn't an explicit goal to remove `timestamp` from the set of built in types?

Reply:
I'm not completely sure about External, but I think they're for a type defined in a different UDL file.  I'm thinking of use-cases like the logins one.  I'm not sure what the plan for `timestamp`, I thought it might be to keep the built-in type, but allow users to override it.

Reply:
>  Isn't an explicit goal to remove timestamp

FTR, I don't think it's necessarily an explicit goal - it would be nice, but I'm not sure it will be possible to replace them without breaking code. My view of the goal is something like "it would have been possible use this PR to get the same functionality without introducing a new type" or something similar.